diff --git a/external/ScrcpyNet/src/ScrcpyNet/ControlMessage.cs b/external/ScrcpyNet/src/ScrcpyNet/ControlMessage.cs
--- a/external/ScrcpyNet/src/ScrcpyNet/ControlMessage.cs
+++ b/external/ScrcpyNet/src/ScrcpyNet/ControlMessage.cs
@@ -127,21 +127,48 @@
     public class ScrollEventControlMessage : IControlMessage
     {
         public ControlMessageType Type => ControlMessageType.InjectScrollEvent;
         public Position Position { get; set; } = new();
         public int HorizontalScroll { get; set; }
         public int VerticalScroll { get; set; }
+        public AndroidMotionEventButtons Buttons { get; set; }
 
         public Span<byte> ToBytes()
         {
-            Span<byte> b = new byte[21];
+            Span<byte> b = new byte[25];
             b[0] = (byte)Type;
             Position.ToBytes().CopyTo(b[1..]);
             BinaryPrimitives.WriteInt32BigEndian(b[13..], HorizontalScroll);
             BinaryPrimitives.WriteInt32BigEndian(b[17..], VerticalScroll);
+            BinaryPrimitives.WriteInt32BigEndian(b[21..], (int)Buttons);
             return b;
         }
     }
+
+    public class RotateDeviceControlMessage : IControlMessage
+    {
+        public ControlMessageType Type => ControlMessageType.RotateDevice;
+
+        public Span<byte> ToBytes()
+        {
+            Span<byte> b = new byte[1];
+            b[0] = (byte)Type;
+            return b;
+        }
+    }
+
+    public class SetScreenPowerModeControlMessage : IControlMessage
+    {
+        public ControlMessageType Type => ControlMessageType.SetScreenPowerMode;
+        public AndroidScreenPowerMode Mode { get; set; } = AndroidScreenPowerMode.POWER_MODE_NORMAL;
+
+        public Span<byte> ToBytes()
+        {
+            Span<byte> b = new byte[2];
+            b[0] = (byte)Type;
+            b[1] = (byte)Mode;
+            return b;
+        }
+    }
 }
